### ISSUE TITLE
Training with LoRA modules

### ISSUE THREAD

**Issue by Abhranta** [ultralytics/ultralytics#8728]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, I want to free the entire yolo model and finetune it using PEFT methods like LoRA. I am able to add the LoRA modules but I am unable to train them. The LoRA modules don't get updated when I call model.train(). The train function here. is the ultralytics trainer. What am I doing wrong?

**Comment by glenn-jocher** [comment]
@Abhranta hi there! 👋 It sounds like you're doing some advanced fine-tuning with LoRA modules on YOLOv8 - that's exciting! If the LoRA modules aren't updating during training, it might be due to how the parameters are set for optimization. Ensure that the parameters of the LoRA modules are explicitly passed to the optimizer. Here's a quick example to check and include LoRA parameters:

```python
from ultralytics import YOLO
import torch

# Load your model
model = YOLO('path/to/your/model.yaml').load('path/to/weights.pt')

# Assuming you've added LoRA modules to 'model' and named them with 'lora' in their names
lora_params = [p for n, p in model.named_parameters() if 'lora' in n]

# Combine LoRA parameters with the rest of the model parameters, excluding those not requiring gradients
params = [p for p in model.parameters() if p.requires_grad] + lora_params

# Setup the optimizer with the combined parameters
optimizer = torch.optim.Adam(params, lr=0.001)

# Now proceed with your training loop
```

This ensures that the optimizer is aware of the LoRA parameters. If they're still not updating, double-check that gradients are not being inadvertently disabled somewhere in your training loop. Hope this helps! Let us know how it goes. 😊

**Comment by Abhranta** [comment]
hey, my question is, can we then user model.train() with the defalt training function of yolo models for this ?

**Comment by glenn-jocher** [comment]
@Abhranta hey there! 😊 Yes, you can use `model.train()` with the default training function of YOLO models, even after integrating LoRA modules. Just ensure your LoRA parameters are properly included in the optimizer's parameter list as shown in the previous example. This way, they'll be updated during the training process. Happy coding!

**Comment by glenn-jocher** [comment]
Hi @Abhranta! 😊

You're correct that our current `model.train()` method primarily allows specifying the optimizer by name, like 'SGD' or 'Adam'. However, for advanced use cases like yours where you want to use custom optimizers or specially prepared parameters (like with LoRA modules), you would currently handle the training loop manually outside of `model.train()` to utilize your custom configured optimizer.

Here's a simplified example of how you can approach training with a custom optimizer in a custom loop:

```python
for epoch in range(num_epochs):
    model.to_train()  # Set the model to training mode
    for batch in dataloader:
        optimizer.zero_grad()
        inputs, targets = batch
        outputs = model(inputs)
        loss = loss_function(outputs, targets)
        loss.backward()
        optimizer.step()
```
This way, you have full control over the optimizer and the training process, allowing detailed customization as per your requirements. 

Feel free to reach out if you need more assistance! 🚀

**Comment by XiphosF** [comment]
Hi @Abhranta , i'm trying to implement LoRA layer in Yolov8 for finetuning on small dataset. I've put LoRA in the convolution block of the detection head, and all seems to work (the LoRA parameters are learning), but it isn't obtaining a good score at the end. Did you manage to obtain something interesting with LoRA in Yolov8 ?

**Comment by yfllllll** [comment]
@XiphosF have you improve it

**Comment by XiphosF** [comment]
@yfllllll tried a few things but never got better metrics with LoRA unfortunately.

**Comment by yfllllll** [comment]
OK，thank you
---- Replied Message ----
***@***.***>Date01/11/2025 17:39 ***@***.***> ***@***.***>,
***@***.***>SubjectRe: [ultralytics/ultralytics] Training with LoRA modules (Issue #8728)
@yfllllll tried a few things but never got better metrics with LoRA unfortunately.

**Comment by glenn-jocher** [comment]
You're welcome! If you have further questions or need additional assistance, feel free to ask. Best of luck with your experiments!

**Comment by Nabyssache** [comment]
yolov8+lora 微调小数据集，nouse
